### PR TITLE
feat(leads): add lead filter and connect it to CRM

### PR DESCRIPTION
**Pull request by anette1983**
## Describe your changes

- I added lead filter and connected it to CRM. User can filter leads by the selected property, by the lead name and by the lead status in different combinations.

**Important notice**: I left console.logs for the development purposes.

image
image
image
image
image

## Issue ticket code (and/or) and link

- [Link to JIRA ticket](https://testhomework4.atlassian.net/jira/software/projects/AG/boards/6?selectedIssue=AG-186)

### **General**

- [x] Assigned myself to the PR
- [x] Assigned the appropriate labels to the PR
- [x] Assigned the appropriate reviewers to the PR
- [x] Updated the documentation
- [x] Performed a self-review of my code
- [x] Types for input and output parameters
- [x] Don't have "any" on my code
- [x] Used the try/catch pattern for error handling
- [x] Don't have magic numbers
- [x] Compare only with constants not with strings
- [x] No ternary operator inside the ternary operator
- [x] Don't have commented code
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [x] Used camelCase for variables and functions
- [ ] Date and time formats are on the constants
- [ ] Functions are public only if it's used outside the class
- [ ] No hardcoded values
- [ ] covered by tests
- [x] Check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Frontend

- [x] Components and business logic are separated
- [x] Colors, Font Size, and Font Name is on the theme or in the constants
- [x] No text in the components, use i18n approach
- [x] No inline styles
- [x] Imports are absolute
- [x] Attach a screenshot if PR has visual changes.

### CODE REVIEW

File: src/components/LeadsFilter/LeadsFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  selectedPropertyId?: string;
+};
+
+type FormValues = {
+  status: string;
+  leadName: string;
+};
+
+export default function LeadsFilter({
+  // selectedPropertyId = '006251f6-39d4-4944-bb2f-c411e0b8d98b',
```
Comment (canterbery) [inline src/components/LeadsFilter/LeadsFilter.tsx]: Commented code

File: src/constants/leadStatuses.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  NEW: { label: 'New', id: 'b041e847-a905-47c1-9f79-430e2fbd6db6' },
+  ASSIGNED: { label: 'Assigned', id: '97e6703e-6282-41d8-91d9-8a911ed6b321' },
+  CONTACT_MADE: { label: 'Contact made', id: '3f6a4fc9-1378-4fca-ad91-894090a6dc72' },
+  VIEWING: { label: 'Viewing', id: '58807212-5e4a-424e-8f69-ece771f03f01' },
+  PROPOSAL: { label: 'Proposal', id: '36760639-bf6d-4dcd-8d9d-25fab4f39254' },
+  NEGOTIATION: { label: 'Negotiation', id: '58ad135b-5036-415b-8ef2-38b417e8ffb4' },
+  ASLEEP: { label: 'Asleep', id: '64622f10-f4ba-4aa1-afb0-73b2571ccc77' },
+  CLOSED_WON: { label: 'Closed won', id: '6ec12111-6ed8-4bed-ac80-4e49e9cefb8b' },
+  CLOSED_LOST: { label: 'Closed lost', id: 'a7ebcea2-975a-4b34-93e7-59eb161c53bc' },
+};
```
Comment (canterbery) [inline src/constants/leadStatuses.ts:11]: why did you hardcode it ?

Cant you just get a list from api and transform it into select options ?

File: src/components/LeadsFilter/LeadsFilter.tsx (hunk 1 above)
Comment (anette1983) [reply src/components/LeadsFilter/LeadsFilter.tsx]: Vyacheslav, I left this code for Sergii to test the filter functionality

File: src/constants/leadStatuses.ts (hunk 2 above)
Comment (anette1983) [reply src/constants/leadStatuses.ts:11]: Unfortunately, I haven't found the way to get the list from the api, also I need this constants to work with filter properly. And it is better for user experience

Comment (canterbery) [reply src/constants/leadStatuses.ts:11]: https://trial5251.eu1.qobrix.com/api/v2/workflows/72c670a3-0521-4b6f-8ce8-6c8299e1dad0/stages

File: src/components/LeadsFilter/LeadsFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  const t = useTranslations('leads');
+
+  const [filterLeads] = useFilterLeadsMutation();
+  const [filterLeadsByProperty] = useFilterLeadsByPropertyMutation();
+
+  const form = useForm<FormValues>({
+    defaultValues: {
+      status: '',
+      leadName: '',
+    },
```
Comment (AKutusenko) [inline src/components/LeadsFilter/LeadsFilter.tsx]: you can take default values out of the component and avoid recreating them every rerender

File: src/components/LeadsFilter/LeadsFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    mode: 'onTouched',
+  });
+  const { register, control, handleSubmit } = form;
+
+  const onSubmit = async (data: FormValues): Promise<void> => {
+    const { status, leadName } = data;
+    const searchString = `ContactNameContacts.name contains '${leadName}' and Opportunities.conversion_status matches '${status}'`;
+    if (selectedPropertyId) {
+      const res = await filterLeadsByProperty({
+        search: searchString,
```
Comment (AKutusenko) [inline src/components/LeadsFilter/LeadsFilter.tsx]: missed try/catch

File: src/components/LeadsFilter/LeadsFilter.tsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      if ('data' in res) {
+        const mappedData = res.data.data.map((item) => ({
+          contact_name: item.contact_name,
+          status: item.status,
+          name: item.contact_name_contact ? item.contact_name_contact.name : null,
+        }));
+        console.log('res with property id', mappedData);
+      }
+    } else {
+      const res = await filterLeads({ conversion_status: status, search: searchString });
```
Comment (AKutusenko) [inline src/components/LeadsFilter/LeadsFilter.tsx]: missed try/catch

File: src/store/auth/authApi.ts [hunk 6]
```diff
@@ -24,7 +24,7 @@ export const authApi = createApi({
         try {
           await queryFulfilled;
         } catch (err) {
-          if (err.error.data.statusCode === 401) {
+          if (err.error.data?.statusCode === 401) {
```
Comment (AKutusenko) [inline src/store/auth/authApi.ts]: you have to compare with constants

**Review by AKutusenko — APPROVED** [review]

File: src/constants/leadStatuses.ts (hunk 2 above)
Comment (lyudad) [reply src/constants/leadStatuses.ts:11]: its not an option at all, when we will change the API key, all will failed as new CRM will have own ip

File: src/components/LeadsFilter/LeadsFilter.tsx (hunk 3 above)
Comment (lyudad) [reply src/components/LeadsFilter/LeadsFilter.tsx]: this should be checked, I think as its hook useForm it should not recreating

File: src/constants/leadStatuses.ts (hunk 2 above)
Comment (anette1983) [reply src/constants/leadStatuses.ts:11]: Lyudmyla, it's not api id, it's a crm lead statuses ids. They won't change, they are always the same for all users

File: src/components/LeadsFilter/LeadsFilter.tsx (hunk 3 above)
Comment (anette1983) [reply src/components/LeadsFilter/LeadsFilter.tsx]: fixed

File: src/components/LeadsFilter/LeadsFilter.tsx (hunk 4 above)
Comment (anette1983) [reply src/components/LeadsFilter/LeadsFilter.tsx]: fixed

File: src/components/LeadsFilter/LeadsFilter.tsx (hunk 5 above)
Comment (anette1983) [reply src/components/LeadsFilter/LeadsFilter.tsx]: it's the same function. Fixed

File: src/store/auth/authApi.ts (hunk 6 above)
Comment (anette1983) [reply src/store/auth/authApi.ts]: fixed

File: src/app/[locale]/leads/page.tsx [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import LeadsFilter from 'components/LeadsFilter/LeadsFilter';
+
+function Leads(): JSX.Element {
+  return (
+    <>
+      <div>Leads</div>
```
Comment (lyudad) [inline src/app/[locale]/leads/page.tsx]: https://www.notion.so/Git-rules-849f8ca8b3db4ded926d762c6fff8798 1.3.3

Comment (anette1983) [reply src/app/[locale]/leads/page.tsx]: Lyudmila, I fixed it, but I want you to know that this is just a sceleton for development purposes, we already have a normal leads page in Sergii's PR. Also, this filter is already integrated in Sergii's PR. I was not going to merge it even, just wanted to get it approved and close for my task to be finished.

Comment (anette1983) [reply src/app/[locale]/leads/page.tsx]: Also, I didn't understand your previous comment ("this should be checked, I think as its hook useForm it should not recreating"), but I can't answer on it directly, don't know why. So, I'll answer here. Everything is working fine. And worked fine before. You want me to transfer default values back to the filter?

Comment (anette1983) [reply src/app/[locale]/leads/page.tsx]: Also, I didn't understand your previous comment ("this should be checked, I think as its hook useForm it should not recreating"), but I can't answer on it directly, don't know why. So, I'll answer here. Everything is working fine. And worked fine before. You want me to transfer default values back to the filter?

File: src/constants/leadStatuses.ts (hunk 2 above)
Comment (lyudad) [reply src/constants/leadStatuses.ts:11]: I'm not talking about users. I mean when we change CRM API key, they will create other instance of db for us = other db = other id

**Review by lyudad — APPROVED** [review]

**Pull request closed without merging**